tachyon/math/polynomials/univariate/radix2_evaluation_domain.h

```diff
@@ -73,6 +85,66 @@ class Radix2EvaluationDomain : public UnivariateEvaluationDomain<F, MaxDegree> {
     return base::bits::SafeLog2Ceiling(num_coeffs) <= F::Config::kTwoAdicity;
   }
 
+  void FFTBatch(RowMajorMatrix<F>& mat) {
+    CHECK(base::bits::IsPowerOfTwo(mat.rows()));
+    size_t log_h = this->log_size_of_group_;
+    mid_ = log_h / 2;
+
+    // The first half looks like a normal DIT.
+    ReverseMatrixIndexBits(mat);
+    RunParallelRowChunks(mat, roots_vec_[0]);
+
+    // For the second half, we flip the DIT, working in bit-reversed order.
+    ReverseMatrixIndexBits(mat);
+    RunParallelRowChunksReversed(mat, roots_vec_[1]);
+    ReverseMatrixIndexBits(mat);
+  }
+
+  CONSTEXPR_IF_NOT_OPENMP void CosetLDEBatch(RowMajorMatrix<F>& mat,
+                                             size_t added_bits) {
+    CHECK(base::bits::IsPowerOfTwo(mat.rows()));
+    size_t log_h = this->log_size_of_group_;
+    mid_ = log_h / 2;
+
+    // The first half looks like a normal DIT.
+    ReverseMatrixIndexBits(mat);
+    RunParallelRowChunks(mat, inv_roots_vec_[0]);
+
+    // For the second half, we flip the DIT, working in bit-reversed order.
+    ReverseMatrixIndexBits(mat);
+    RunParallelRowChunksReversed(mat, inv_roots_vec_[1]);
+    // We skip the final bit-reversal, since the next FFT expects bit-reversed
+    // input.
+
+    // Rescale coefficients in two ways:
+    // - divide by number of rows (since we're doing an inverse DFT)
+    // - multiply by powers of the coset shift (see default coset LDE impl for
+    std::vector<F> weights = F::GetSuccessivePowers(
+        this->size_, F::FromMontgomery(F::Config::kSubgroupGenerator),
+        this->size_inv_);
+    OPENMP_PARALLEL_FOR(size_t row = 0; row < weights.size(); ++row) {
+      // Reverse bits because mat is encoded in bit-reversed order
+      mat.row(base::bits::BitRev(row) >>
+              (sizeof(size_t) * 8 - this->log_size_of_group_)) *= weights[row];
+    }
+    ExpandInPlaceWithZeroPad<RowMajorMatrix<F>>(mat, added_bits);
+
+    size_t rows = static_cast<size_t>(mat.rows());
+    CHECK(base::bits::IsPowerOfTwo(rows));
```

Comment:
It's a redundant check. You're already checking this at the first line of this function.

Reply:
The size of `mat` and therefore the number of rows changed from `ExpandInPlaceWithZeroPad`, so we need to check again!